tests/test_steps.py

```diff
@@ -344,6 +344,23 @@ def test_sample_emit_last_with_emit_before_termination():
     controller.await_termination()
 
 
+def test_sample_emit_event_per_key():
+    controller = build_flow(
+        [
+            SyncEmitSource(key_field=str),
+            Assert().exactly(25),
+            SampleWindow(5, key='$key'),
+            Assert().exactly(5).match_exactly([0, 1, 2, 3, 4]),
```

Comment:
Don't use `Assert`. It won't fail if nothing is ever emitted by `SampleWindow`. Use `Reduce` and `assert` instead.

Reply:
Never mind, I see this is used all over in this test...